phoenix-core-client/src/main/java/org/apache/phoenix/util/CDCChangeBuilder.java

```diff
@@ -0,0 +1,151 @@
+    private String changeType;
+    private long lastDeletedTimestamp;
+    private long changeTimestamp;
+    private Map<String, Object> preImage = null;
+    private Map<String, Object> changeImage = null;
+
+    public CDCChangeBuilder(CDCTableInfo cdcDataTableInfo) {
+        this.cdcDataTableInfo = cdcDataTableInfo;
+        Set<PTable.CDCChangeScope> changeScopes = cdcDataTableInfo.getIncludeScopes();
+        isChangeImageInScope = changeScopes.contains(PTable.CDCChangeScope.CHANGE);
+        isPreImageInScope = changeScopes.contains(PTable.CDCChangeScope.PRE);
+        isPostImageInScope = changeScopes.contains(PTable.CDCChangeScope.POST);
+    }
+
+    public void initChange(long ts) {
+        changeTimestamp = ts;
+        changeType = null;
+        lastDeletedTimestamp = 0L;
+        if (isPreImageInScope || isPostImageInScope) {
+            preImage = new HashMap<>();
+        }
+        if (isChangeImageInScope || isPostImageInScope) {
+            changeImage = new HashMap<>();
+        }
+    }
+
+    public long getChangeTimestamp() {
+        return changeTimestamp;
+    }
+
+    public boolean isDeletionEvent() {
+        return changeType == CDC_DELETE_EVENT_TYPE;
+    }
+
+    public boolean isNonEmptyEvent() {
+        return changeType != null;
+    }
+
+    public void markAsDeletionEvent() {
+        changeType = CDC_DELETE_EVENT_TYPE;
+    }
+
+    public long getLastDeletedTimestamp() {
+        return lastDeletedTimestamp;
+    }
+
+    public void setLastDeletedTimestamp(long lastDeletedTimestamp) {
+        this.lastDeletedTimestamp = lastDeletedTimestamp;
+    }
+
+    public boolean isChangeRelevant(Cell cell) {
+        if (cell.getTimestamp() > changeTimestamp) {
+            return false;
+        }
+        if (cell.getType() != Cell.Type.DeleteFamily && !isOlderThanChange(cell) &&
+                isDeletionEvent()) {
+            // We don't need to build the change image in this case.
+            return false;
+        }
+        return true;
+    }
+
+    public void registerChange(Cell cell, int columnNum, Object value) {
+        if (!isChangeRelevant(cell)) {
+            return;
+        }
+        CDCTableInfo.CDCColumnInfo columnInfo =
+                cdcDataTableInfo.getColumnInfoList().get(columnNum);
+        String cdcColumnName = columnInfo.getColumnDisplayName(cdcDataTableInfo);
+        if (isOlderThanChange(cell)) {
+            if ((isPreImageInScope || isPostImageInScope) &&
+                    !preImage.containsKey(cdcColumnName)) {
+                preImage.put(cdcColumnName, value);
+            }
+        } else if (cell.getTimestamp() == changeTimestamp) {
+            assert !isDeletionEvent() : "Not expected to find a change for delete event";
+            changeType = CDC_UPSERT_EVENT_TYPE;
+            if (isChangeImageInScope || isPostImageInScope) {
+                changeImage.put(cdcColumnName, value);
+            }
+        }
+    }
+
+    public Map buildCDCEvent() {
```

Comment:
All the places where raw `Map` is used, let's replace it with `Map<String, Object>`